File: src/main/java/com/massivecraft/massivecore/util/reference/ProviderOptimizedCollectionSafe.java
package com.massivecraft.massivecore.util.reference;

import java.util.EnumSet;
import java.util.HashMap;
import java.util.Map;

public class ProviderOptimizedCollectionSafe {

    // For building an enum set based on a Class and names of the constants, but just skip elements that don't match
    public static <E extends Enum<E>> EnumSet<E> enumSetOf(Class<E> clazz, String... names) {
        Map<String, E> namesToElements = getNameToElementForEnum(clazz);
        EnumSet<E> ret = EnumSet.noneOf(clazz);
        for (String name : names) {
            E element = namesToElements.get(name);
            if (element != null) {
                ret.add(element);
            }
        }
        return ret;
    }

    // Creating this map can end up being much more efficient when we have to process large sets of names or operate on large enums
    // It prevents repeated iteration
    private static <E extends Enum<E>> Map<String, E> getNameToElementForEnum(Class<E> clazz) {
        E[] enumConstants = clazz.getEnumConstants();
        if (enumConstants == null) throw new IllegalArgumentException("clazz is not an Enum");
        Map<String, E> ret = new HashMap<>();
        for (E element : enumConstants) {
            ret.put(element.name(), element);
        }
        return ret;
    }
}

File: src/com/massivecraft/massivecore/util/reference/ProviderOptimizedCollectionSafe.kt
package com.massivecraft.massivecore.util.reference

import java.util.EnumSet

// MassiveTraits
// TODO TraitEatCarnivore
// TODO TraitEatHerbivore
// TODO TraitEatStone
// TODO TraitEatWaterBottle
// TODO Trait
// TODO TraitHarvester
// TODO TraitTruceUndead
// TODO TraitTruceSpider
// TODO TraitTruceWitch
// TODO TraitTruceGhast
// TODO TraitTruceSilme
// TODO TraitTruceSilverfish
// TODO TraitTruceUndead
// TODO TraitTruceCreeper
// TODO TraitTruceBlaze

// MassiveCore
// TODO MUtil: HARMFUL_POTION_EFFECTS
// TODO SignUtil: VALID_INTERACT_ACTIONS, isSign(Material material)
// TODO
// TODO
// https://docs.google.com/document/d/1m4Nzhgo6CdiTxVzu1gqVGyQ24xsLkSpHqK8yxb33kuw/edit
object ProviderOptimizedCollectionSafe {

    fun <E: Enum<E>> enumSetOf(clazz: Class<E>, vararg names: String): EnumSet<E> {
        val mapper = getEnumMapper(clazz)
        val ret = EnumSet.noneOf(clazz)
        return names.mapNotNullTo(ret, mapper)
    }

    // TODO memoize
    private fun <E: Enum<E>> getEnumMapper(clazz: Class<E>): DestringifierSafe<E> {
        val mapping = clazz.enumConstants.associateBy { constant -> constant.name }
        return fun(name: String): E? {
            return mapping[name]
        }
    }



    /*private inline fun <E: Any, S: Set<E>> setBuilderOptimized(
            mapper: (String) -> E?,
            setConstructor: (Iterable<E>) -> S,
            names: Iterable<String>
    ): S {
        return setConstructor.invoke(names.mapNotNull(mapper))
    }*/

}

typealias DestringifierSafe<E> = (String) -> E?
